File: src/services/StatsService.ts
import { db } from "../config/database";
import {
  DashboardStats,
  DailyStats,
  TypeDistribution,
  GeographicDistribution,
} from "../types/stats";

export class StatsService {
  async getDashboardStats(): Promise<DashboardStats> {
    // Statistiques globales
    const globalStats = await db.query(`
            SELECT 
                COUNT(DISTINCT e.id) as total_epidemics,
                COUNT(DISTINCT CASE WHEN e.end_date IS NULL THEN e.id END) as active_epidemics,
                SUM(ds.cases) as total_cases,
                SUM(ds.deaths) as total_deaths,
                ROUND(CAST(SUM(ds.deaths) AS FLOAT) / NULLIF(SUM(ds.cases), 0) * 100, 2) as mortality_rate
            FROM Epidemic e
            LEFT JOIN Daily_stats ds ON e.id = ds.id_epidemic
            WHERE ds.date = (
                SELECT MAX(date) 
                FROM Daily_stats 
                WHERE id_epidemic = e.id
            )
        `);

    // Distribution par type
    const typeDistribution = await db.query(`
            SELECT 
                e.type,
                COUNT(DISTINCT e.id) as epidemic_count,
                SUM(ds.cases) as cases,
                SUM(ds.deaths) as deaths
            FROM Epidemic e
            LEFT JOIN Daily_stats ds ON e.id = ds.id_epidemic
            WHERE ds.date = (
                SELECT MAX(date) 
                FROM Daily_stats 
                WHERE id_epidemic = e.id
            )
            GROUP BY e.type
            ORDER BY cases DESC
        `);

    // Distribution géographique
    const geographicDistribution = await db.query(`
            SELECT 
                l.country,
                SUM(ds.cases) as cases,
                SUM(ds.deaths) as deaths
            FROM Daily_stats ds
            JOIN Localisation l ON ds.id_loc = l.id
            WHERE ds.date = (
                SELECT MAX(date) 
                FROM Daily_stats 
                WHERE id_epidemic = ds.id_epidemic
            )
            GROUP BY l.country
            ORDER BY cases DESC
            LIMIT 10
        `);

    // Évolution quotidienne (30 derniers jours)
    const dailyEvolution = await db.query(`
            SELECT 
                ds.date,
                SUM(ds.new_cases) as new_cases,
                SUM(ds.new_deaths) as new_deaths,
                SUM(ds.active) as active_cases
            FROM Daily_stats ds
            WHERE ds.date >= DATE_SUB(CURDATE(), INTERVAL 30 DAY)
            GROUP BY ds.date
            ORDER BY ds.date
        `);

    // Top épidémies actives
    const topActiveEpidemics = await db.query(`
            SELECT 
                e.id,
                e.name,
                e.type,
                l.country,
                ds.cases as total_cases,
                ds.deaths as total_deaths
            FROM Epidemic e
            JOIN Daily_stats ds ON e.id = ds.id_epidemic
            JOIN Localisation l ON ds.id_loc = l.id
            WHERE e.end_date IS NULL
            AND ds.date = (
                SELECT MAX(date) 
                FROM Daily_stats 
                WHERE id_epidemic = e.id
            )
            ORDER BY ds.cases DESC
            LIMIT 5
        `);

    return {
      global_stats: {
        total_epidemics: globalStats[0].total_epidemics,
        active_epidemics: globalStats[0].active_epidemics,
        total_cases: globalStats[0].total_cases,
        total_deaths: globalStats[0].total_deaths,
        mortality_rate: globalStats[0].mortality_rate,
      },
      type_distribution: typeDistribution.map((type) => ({
        type: type.type,
        epidemic_count: type.epidemic_count,
        cases: type.cases,
        deaths: type.deaths,
      })),
      geographic_distribution: geographicDistribution.map((geo) => ({
        country: geo.country,
        cases: geo.cases,
        deaths: geo.deaths,
      })),
      daily_evolution: dailyEvolution.map((day) => ({
        date: day.date,
        new_cases: day.new_cases,
        new_deaths: day.new_deaths,
        active_cases: day.active_cases,
      })),
      top_active_epidemics: topActiveEpidemics.map((epidemic) => ({
        id: epidemic.id,
        name: epidemic.name,
        type: epidemic.type,
        country: epidemic.country,
        total_cases: epidemic.total_cases,
        total_deaths: epidemic.total_deaths,
      })),
    };
  }

  async getDailyStats(epidemicId?: number): Promise<DailyStats[]> {
    const query = `
            SELECT 
                ds.id,
                ds.date,
                ds.cases,
                ds.active,
                ds.deaths,
                ds.recovered,
                ds.new_cases,
                ds.new_deaths,
                ds.new_recovered,
                e.id as epidemic_id,
                e.name as epidemic_name,
                e.type as epidemic_type,
                l.country,
                l.region
            FROM Daily_stats ds
            JOIN Epidemic e ON ds.id_epidemic = e.id
            JOIN Localisation l ON ds.id_loc = l.id
            ${epidemicId ? "WHERE ds.id_epidemic = ?" : ""}
            ORDER BY ds.date DESC
            LIMIT 100
        `;

    const params = epidemicId ? [epidemicId] : [];
    const results = await db.query(query, params);

    return results.map((row) => ({
      id: row.id,
      date: row.date,
      cases: row.cases,
      active: row.active,
      deaths: row.deaths,
      recovered: row.recovered,
      new_cases: row.new_cases,
      new_deaths: row.new_deaths,
      new_recovered: row.new_recovered,
      epidemic: {
        id: row.epidemic_id,
        name: row.epidemic_name,
        type: row.epidemic_type,
      },
      location: {
        country: row.country,
        region: row.region,
      },
    }));
  }

  async getTypeStats(): Promise<TypeDistribution[]> {
    const results = await db.query(`
            SELECT 
                e.type,
                COUNT(DISTINCT e.id) as count,
                SUM(ds.cases) as total_cases,
                SUM(ds.deaths) as total_deaths,
                ROUND(AVG(ds.active), 0) as avg_active_cases
            FROM Epidemic e
            LEFT JOIN Daily_stats ds ON e.id = ds.id_epidemic
            WHERE ds.date = (
                SELECT MAX(date) 
                FROM Daily_stats 
                WHERE id_epidemic = e.id
            )
            GROUP BY e.type
            ORDER BY total_cases DESC
        `);

    return results.map((row) => ({
      type: row.type,
      count: row.count,
      total_cases: row.total_cases,
      total_deaths: row.total_deaths,
      avg_active_cases: row.avg_active_cases,
    }));
  }

  async getGeographicStats(): Promise<GeographicDistribution[]> {
    const results = await db.query(`
            SELECT 
                l.country,
                l.region,
                COUNT(DISTINCT e.id) as epidemic_count,
                SUM(ds.cases) as total_cases,
                SUM(ds.deaths) as total_deaths,
                ROUND(AVG(ds.active), 0) as avg_active_cases
            FROM Localisation l
            JOIN Daily_stats ds ON l.id = ds.id_loc
            JOIN Epidemic e ON ds.id_epidemic = e.id
            WHERE ds.date = (
                SELECT MAX(date) 
                FROM Daily_stats 
                WHERE id_epidemic = e.id
            )
            GROUP BY l.country, l.region
            ORDER BY total_cases DESC
        `);

    return results.map((row) => ({
      country: row.country,
      region: row.region || null,
      epidemic_count: row.epidemic_count,
      total_cases: row.total_cases,
      total_deaths: row.total_deaths,
      avg_active_cases: row.avg_active_cases,
    }));
  }
}

File: app/services/stats_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func, distinct, desc
from typing import List, Dict, Any
from ..db.models.base import Epidemic, DailyStats, Localisation

class StatsService:
    def __init__(self, db: Session):
        self.db = db

    def get_dashboard_stats(self) -> Dict[str, Any]:
        """
        Récupère toutes les statistiques pour le tableau de bord
        """
        return {
            "global_stats": self._get_global_stats(),
            "type_distribution": self._get_type_distribution(),
            "geographic_distribution": self._get_geographic_distribution(),
            "daily_evolution": self._get_daily_evolution(),
            "top_active_epidemics": self._get_top_active_epidemics()
        }

    def _get_global_stats(self) -> Dict[str, Any]:
        """
        Calcule les statistiques globales
        """
        # Récupérer les dernières statistiques quotidiennes
        latest_stats = self.db.query(
            func.sum(DailyStats.cases).label('total_cases'),
            func.sum(DailyStats.deaths).label('total_deaths')
        ).first()

        # Compter les épidémies
        total_epidemics = self.db.query(func.count(distinct(DailyStats.id_epidemic))).scalar()
        
        # Calculer le taux de mortalité
        total_cases = int(latest_stats.total_cases or 0)
        total_deaths = int(latest_stats.total_deaths or 0)
        mortality_rate = (total_deaths / total_cases * 100) if total_cases > 0 else 0

        return {
            "total_cases": total_cases,
            "total_deaths": total_deaths,
            "total_epidemics": total_epidemics,
            "active_epidemics": total_epidemics,  # Comme nous n'avons pas de flag actif/inactif
            "mortality_rate": float(mortality_rate)
        }

    def _get_type_distribution(self) -> List[Dict[str, Any]]:
        """
        Calcule la distribution par type d'épidémie
        """
        results = self.db.query(
            Epidemic.type,
            func.sum(DailyStats.cases).label('cases'),
            func.sum(DailyStats.deaths).label('deaths')
        ).join(
            DailyStats,
            DailyStats.id_epidemic == Epidemic.id
        ).group_by(Epidemic.type).all()

        return [
            {
                "type": result.type or "Non spécifié",
                "cases": int(result.cases or 0),
                "deaths": int(result.deaths or 0)
            }
            for result in results
        ]

    def _get_geographic_distribution(self) -> List[Dict[str, Any]]:
        """
        Calcule la distribution géographique des cas
        """
        results = self.db.query(
            Localisation.country,
            func.sum(DailyStats.cases).label('cases'),
            func.sum(DailyStats.deaths).label('deaths')
        ).join(
            DailyStats,
            DailyStats.id_loc == Localisation.id
        ).group_by(
            Localisation.country
        ).order_by(
            desc('cases')
        ).limit(10).all()

        return [
            {
                "country": result.country,
                "cases": int(result.cases or 0),
                "deaths": int(result.deaths or 0)
            }
            for result in results
        ]

    def _get_daily_evolution(self) -> List[Dict[str, Any]]:
        """
        Récupère l'évolution quotidienne des cas sur les 30 derniers jours
        """
        # Récupérer toutes les dates disponibles
        dates = self.db.query(
            DailyStats.date
        ).distinct().order_by(DailyStats.date.asc()).all()
        
        if not dates:
            return []

        results = self.db.query(
            DailyStats.date,
            func.sum(DailyStats.new_cases).label('new_cases'),
            func.sum(DailyStats.new_deaths).label('new_deaths'),
            func.sum(DailyStats.active).label('active_cases')
        ).group_by(
            DailyStats.date
        ).order_by(
            DailyStats.date.asc()
        ).all()

        return [
            {
                "date": result.date.isoformat(),
                "new_cases": int(result.new_cases or 0),
                "new_deaths": int(result.new_deaths or 0),
                "active_cases": int(result.active_cases or 0)
            }
            for result in results
        ]

    def _get_top_active_epidemics(self) -> List[Dict[str, Any]]:
        """
        Récupère les épidémies les plus importantes
        """
        # Sous-requête pour obtenir les totaux par épidémie
        epidemic_stats = self.db.query(
            DailyStats.id_epidemic,
            func.sum(DailyStats.cases).label('total_cases'),
            func.sum(DailyStats.deaths).label('total_deaths')
        ).group_by(
            DailyStats.id_epidemic
        ).subquery()

        results = self.db.query(
            Epidemic.id,
            Epidemic.name,
            Epidemic.type,
            Epidemic.country,
            epidemic_stats.c.total_cases,
            epidemic_stats.c.total_deaths
        ).join(
            epidemic_stats,
            epidemic_stats.c.id_epidemic == Epidemic.id
        ).order_by(
            desc(epidemic_stats.c.total_cases)
        ).limit(5).all()

        return [
            {
                "id": result.id,
                "name": result.name,
                "type": result.type or "Non spécifié",
                "country": result.country,
                "total_cases": int(result.total_cases or 0),
                "total_deaths": int(result.total_deaths or 0)
            }
            for result in results
        ] 